libs/constrained-generators/src/Constrained/Instances.hs

```diff
@@ -67,5 +67,25 @@ instance IsUniverse fn => Functions (BoolFn fn) fn where
       let args = appendList (mapList (\(Value a) -> Lit a) pre) (v' :> mapList (\(Value a) -> Lit a) suf)
        in Let (App (injectFn fn) args) (v :-> ps)
   propagateSpecFun Not (NilCtx HOLE) spec = caseBoolSpec spec (equalSpec . not)
+  propagateSpecFun And (HOLE :? Value (s :: Bool) :> Nil) spec = caseBoolSpec spec (okAnd s)
+  propagateSpecFun And (Value (s :: Bool) :! NilCtx HOLE) spec = caseBoolSpec spec (okAnd s)
+  propagateSpecFun Or (HOLE :? Value (s :: Bool) :> Nil) spec = caseBoolSpec spec (okOr s)
+  propagateSpecFun Or (Value (s :: Bool) :! NilCtx HOLE) spec = caseBoolSpec spec (okOr s)
 
   mapTypeSpec Not _ = typeSpec ()
+
+-- | We have something like ('constant' &&. HOLE) must evaluate to 'need'. Return a (Spec fn Bool) for HOLE, that makes that True.
+okAnd :: Bool -> Bool -> Spec fn Bool
```

Comment:
I would put these in a `were` clause to keep the top level interface of the module clean, but that's just a style thing for me.

Reply:
It is hard to get these function in the two scopes where the are called, and still make it clear that propagateSpecFun exhaustively covers all cases, if they are not at the top level. So I plan to leave them where they are, But we could make a more explicit statement of what is exported at the module level. Let me know if you want to do that.

Reply:
No this is ok, no big deal at all.